File: rileylink/src/main/java/info/nightscout/androidaps/plugins/pump/common/hw/medlink/ble/CommandExecutor.java
package info.nightscout.androidaps.plugins.pump.common.hw.medlink.ble;

import androidx.annotation.NonNull;

import java.util.Objects;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Stream;

import info.nightscout.androidaps.logging.AAPSLogger;
import info.nightscout.androidaps.logging.LTag;
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.activities.BolusCallback;
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.activities.BolusProgressCallback;
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.activities.MedLinkStandardReturn;
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.ble.command.BleCommand;
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.ble.data.BolusMedLinkMessage;
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.ble.data.MedLinkPumpMessage;
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.defs.MedLinkCommandType;
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.service.MedLinkServiceData;

/**
 * Created by Dirceu on 01/04/21.
 */
public abstract class CommandExecutor implements Runnable {
    private final MedLinkPumpMessage<?> medLinkPumpMessage;
    private final AAPSLogger aapsLogger;
    private int commandPosition = 0;
    private int functionPosition = 0;
    private MedLinkCommandType currentCommand;

    protected int nrRetries = 0;

    protected CommandExecutor(MedLinkPumpMessage<?> medLinkPumpMessage, AAPSLogger aapsLogger) {
        this.medLinkPumpMessage = medLinkPumpMessage;
        this.aapsLogger = aapsLogger;
    }

    protected CommandExecutor(MedLinkCommandType commandType, AAPSLogger aapsLogger, MedLinkServiceData medLinkServiceData) {
        this(new MedLinkPumpMessage(commandType,
                new BleCommand(aapsLogger, medLinkServiceData)), aapsLogger);
    }

//
//    protected CommandExecutor(RemainingBleCommand remainingBleCommand) {
//        this.remainingBleCommand = remainingBleCommand;
//    }


    public boolean contains(MedLinkCommandType com) {
        return (com != null && com.isSameCommand(medLinkPumpMessage.getCommandType()) ||
                com.isSameCommand(medLinkPumpMessage.getArgument()));
    }

    public MedLinkCommandType nextCommand() {
        if (commandPosition == 0) {
            return medLinkPumpMessage.getCommandType();
        } else if (commandPosition == 1) {
            return medLinkPumpMessage.getArgument();
        } else return MedLinkCommandType.NoCommand;
    }

    public Function<Supplier<Stream<String>>, ? extends MedLinkStandardReturn<?>> nextFunction() {
        if (functionPosition == 0 && medLinkPumpMessage.getBaseCallback() != null) {
            currentCommand = medLinkPumpMessage.getCommandType();
            return medLinkPumpMessage.getBaseCallback().andThen(f -> {
                functionPosition += 1;
                return f;
            });
        } else if (functionPosition == 1 && medLinkPumpMessage.getArgCallback() != null) {
            currentCommand = medLinkPumpMessage.getArgument();
            return medLinkPumpMessage.getArgCallback().andThen(f -> {
                functionPosition += 1;
                return f;
            });
        } else return null;
    }

    public MedLinkCommandType getCurrentCommand() {
        if (commandPosition <= 1) {
            return medLinkPumpMessage.getCommandType();
        } else if (commandPosition == 2) {
            return medLinkPumpMessage.getArgument();
        } else {
            return MedLinkCommandType.NoCommand;
        }
    }

    public boolean hasFinished() {
        aapsLogger.info(LTag.PUMPBTCOMM, medLinkPumpMessage.toString());
        aapsLogger.info(LTag.PUMPBTCOMM, "" + commandPosition);
        return commandPosition > 1
                || (MedLinkCommandType.NoCommand.isSameCommand(
                medLinkPumpMessage.getArgument()) &&
                commandPosition > 0);

    }

    public void commandExecuted() {
        this.commandPosition += 1;
    }

    public void clearExecutedCommand() {
        this.commandPosition = 0;
        this.functionPosition = 0;
    }

    @NonNull @Override public String toString() {
        return "CommandExecutor{" +
                "command=" + medLinkPumpMessage +
                ", commandPosition=" + commandPosition +
                ", functionPosition=" + functionPosition +
                '}';
    }

    public MedLinkPumpMessage<?> getMedLinkPumpMessage() {
        return medLinkPumpMessage;
    }

    public int getNrRetries() {
        return nrRetries;
    }

    public boolean matches(MedLinkPumpMessage<?> ext) {
        return this.getMedLinkPumpMessage().getCommandType().isSameCommand(ext.getCommandType())
                && this.getMedLinkPumpMessage().getArgument().isSameCommand(ext.getArgument());
    }

    @Override public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        CommandExecutor that = (CommandExecutor) o;
        return !hasFinished() && !that.hasFinished() &&
                medLinkPumpMessage.getCommandType() == that.medLinkPumpMessage.getCommandType() &&
                medLinkPumpMessage.getArgument() == that.medLinkPumpMessage.getArgument();
    }

    @Override public int hashCode() {
        return Objects.hash(medLinkPumpMessage, aapsLogger, commandPosition, functionPosition, currentCommand, nrRetries);
    }

    protected byte[] nextCommandData() {
        if (commandPosition == 0) {
            return medLinkPumpMessage.getCommandType().getRaw();
        } else if (commandPosition == 1) {
            return medLinkPumpMessage.getArgumentData();
        } else return MedLinkCommandType.NoCommand.getRaw();
    }
}

File: rileylink/src/main/java/info/nightscout/androidaps/plugins/pump/common/hw/medlink/ble/CommandExecutor.kt
package info.nightscout.androidaps.plugins.pump.common.hw.medlink.ble

import info.nightscout.androidaps.plugins.pump.common.hw.medlink.activities.MedLinkStandardReturn
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.ble.data.MedLinkPumpMessage
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.defs.MedLinkCommandType
import info.nightscout.shared.logging.AAPSLogger
import info.nightscout.shared.logging.LTag
import java.util.*
import java.util.function.Function
import java.util.function.Supplier
import java.util.stream.Stream

/**
 * Created by Dirceu on 01/04/21.
 */
abstract class CommandExecutor protected constructor(
    val commandList: List<
        Pair<
            MedLinkCommandType,
            Function<
                Supplier<Stream<String>>,
                MedLinkStandardReturn<*>
                >?
            >
        >, val aapsLogger: AAPSLogger
) : Runnable {

    private var commandPosition = 0
    private var functionPosition = 0
    private var currentCommand: MedLinkCommandType? = null
    var nrRetries = 0
        protected set

    protected constructor(commandType: MedLinkCommandType, aapsLogger: AAPSLogger) :
        this(
            listOf(
                Pair<
                    MedLinkCommandType,
                    Function<
                        Supplier<Stream<String>>,
                        MedLinkStandardReturn<*>
                        >?
                    >(
                    commandType, null
                )
            ), aapsLogger
        ) {

    }

    //
    //    protected CommandExecutor(RemainingBleCommand remainingBleCommand) {
    //        this.remainingBleCommand = remainingBleCommand;
    //    }
    fun contains(com: MedLinkCommandType?): Boolean {
        return com != null && commandList.any { it.first.isSameCommand(com) }
    }

    fun nextCommand(): MedLinkCommandType {
        return if (commandPosition >= commandList.size) {
            MedLinkCommandType.NoCommand
        } else {
            commandList[commandPosition].first
        }
    }

    fun nextFunction(): Function<Supplier<Stream<String>>, out MedLinkStandardReturn<*>?>? {
        return if (commandList[functionPosition].second != null) {
            currentCommand = commandList[functionPosition].first
            commandList[functionPosition].second?.compose {
                aapsLogger.info(
                    LTag.APS, "applied"
                )
                functionPosition += 1
                it
            }

            // } else if (functionPosition == 1 && medLinkPumpMessage.argCallback != null) {
            //     currentCommand = medLinkPumpMessage.argument
            //     medLinkPumpMessage.argCallback.compose {
            //         functionPosition += 1
            //         it
            //     }
            // } else if (functionPosition == 1 && medLinkPumpMessage is BasalMedLinkMessage) {
            //     medLinkPumpMessage.profileCallback.compose {
            //         functionPosition += 1
            //         it
            //     }
        } else null
    }

    fun getCurrentCommand(): MedLinkCommandType {
        return if (commandPosition < commandList.size)
            commandList[commandPosition].first
        else MedLinkCommandType.NoCommand
    }

    open fun hasFinished(): Boolean {
        aapsLogger.info(LTag.PUMPBTCOMM, commandList.joinToString())
        aapsLogger.info(LTag.PUMPBTCOMM, "" + commandPosition)
        return (commandPosition >= commandList.size
            )
    }

    fun commandExecuted() {
        if (commandPosition == 0) {
            nrRetries++
        } else if (commandList[commandPosition].first != MedLinkCommandType.NoCommand) {
            nrRetries++
        }
        commandPosition += 1
    }

    open fun clearExecutedCommand() {
        commandPosition = 0
        functionPosition = 0
    }

    override fun toString(): String {
        return "CommandExecutor{" +
            "command=" + commandList.joinToString() +
            ", commandPosition=" + commandPosition +
            ", functionPosition=" + functionPosition +
            ", nrRetries=" + nrRetries +
            '}'
    }

    fun matches(ext: MedLinkPumpMessage<*>): Boolean {
        return (commandList == ext.commands)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as CommandExecutor
        return !hasFinished() && !that.hasFinished() && commandList == other.commandList
    }

    override fun hashCode(): Int {
        return Objects.hash(commandList, aapsLogger, commandPosition, functionPosition, currentCommand, nrRetries)
    }

    protected fun nextCommandData(): ByteArray {
        return if (commandPosition < commandList.size) {
            commandList[commandPosition].first.raw
        } else {
            MedLinkCommandType.NoCommand.raw
        }
    }

    val isInitialized: Boolean
        get() = commandPosition > 0 || functionPosition > 0

}